KAFKA-6576: Configurable Quota Management (KIP-257)

rajinisivaram opened the pull request:
Enable quota calculation to be customized using a configurable callback. See KIP-257 for details.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rajinisivaram committed:
KAFKA-6576: Configurable Quota Management (KIP-257)

junrao commented on core/src/main/scala/kafka/quota/ClientQuotaCallback.scala:
```diff
@@ -0,0 +1,139 @@
+
+  /**
+   * Metadata update callback that is invoked whenever UpdateMetadata request is received from
+   * the controller. This is useful if quota computation is takes partitions into account.
```
Typo "is takes ".

rajinisivaram commented:
retest this please

junrao commented on core/src/main/scala/kafka/quota/ClientQuotaCallback.scala:
```diff
@@ -0,0 +1,139 @@
+   *
+   * @return the quota including the limit and metric tags that indicate which other entities share this quota
+   */
+  def quota(session: Session, clientId: String, quotaType: ClientQuotaType): ClientQuota
```
Since we already have quotaLimit(), could quota() just return the metric tags? The caller can then use the metric tags to obtain the quota limit. This way, there is only one way to get the quota limit.

rajinisivaram replied:
@junrao We need to go through the quota configs at different levels to determine the metric tags. `quotaLimit` also does something similar. So it is cheaper to return the limit with the tags, and since this is used for every request, it may be worth returning limit and tags together. With two calls, we also need to handle changes between the two invocations. Alternatively, we could do something different for default callback handler alone since custom handlers are less likely to have several levels. What do you think?

junrao replied:
@rajinisivaram : Hmm, my understanding is that quota() only needs to be called when the quota metric doesn't already exist. So, typically, quota() is only called on the first request from a given client and not on every request.

rajinisivaram replied:
@junrao At the moment, we call `quota()` on every request since need to work out which metric tags to use and to do that we need to know which level of configs are defined. But, you are right, after the first request, we already know the limit since the metric exists, so we don't need to call two methods every time. So agree, we can change the first method to just return metric tags. Will update the PR, thanks.

rajinisivaram replied:
@junrao I have changed the first method to `quotaMetricTags` to obtain just the tags and updated the code to use `quotaLimit` to get the limit in all cases. Thanks.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -421,86 +379,238 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
       val singleUpdate = quotaTypesEnabled match {
         case QuotaTypes.NoQuotas | QuotaTypes.ClientIdQuotaEnabled | QuotaTypes.UserQuotaEnabled | QuotaTypes.UserClientIdQuotaEnabled =>
-          !sanitizedUser.filter(_ == ConfigEntityName.Default).isDefined && !clientId.filter(_ == ConfigEntityName.Default).isDefined
+          quotaEntity.userEntity != Some(DefaultUserEntity) && quotaEntity.clientIdEntity != Some(DefaultClientIdEntity)
```
The comment in line 472 probably needs to be adjusted.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -421,86 +379,238 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
-            info(s"Sensor for ${userInfo}${clientIdInfo} already exists. Changing quota to ${newQuota.bound()} in MetricConfig")
-            metric.config(getQuotaMetricConfig(newQuota))
+        // Change the underlying metric config if the sensor has been created
+        val metric = allMetrics.get(quotaMetricName)
```
Ideally, the callback shouldn't access internal objects such as allMetrics. So perhaps the logic under singleUpdate should be handled in the same place where multiple updates are handled?

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -188,8 +243,8 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
-        val clientMetric = metrics.metrics().get(clientRateMetricName(clientQuotaEntity.sanitizedUser, clientQuotaEntity.clientId))
-        throttleTimeMs = throttleTime(clientMetric, getQuotaMetricConfig(clientQuotaEntity.quota)).toInt
+        val clientMetric = metrics.metrics().get(clientRateMetricName(clientQuotaEntity.metricTags))
+        throttleTimeMs = throttleTime(clientMetric, getQuotaMetricConfig(clientQuotaEntity.quotaLimit)).toInt
```
This is not introduced in this patch. But I am wondering why we need to pass in a separate MetricConfig to throttleTime() since clientMetric contains MetricConfig already.

rajinisivaram replied:
Not sure, I changed it and tests still passed, so I guess it is to remove it.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -360,56 +312,62 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
+  private def getQuotaSensorName(quotaEntity: ClientQuota): String = s"$quotaType-${quotaIdToSensorSuffix(quotaEntity)}"
 
-  private def getQuotaSensorName(quotaId: QuotaId): String = quotaType + "-" + quotaId.sanitizedUser.getOrElse("") + ':' + quotaId.clientId.getOrElse("")
+  private def metricTags(quotaEntity: ClientQuota): Map[String, String] = quotaEntity.metricTags
```
It seems metricTags is never used? Not sure if it's worth having a method to just get one field.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -421,86 +379,238 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
-        case None => ""
+      val clientIdEntity = sanitizedClientId.map {
+        case ConfigEntityName.Default => DefaultClientIdEntity
+        case id => ClientIdEntity(clientId.getOrElse(throw new IllegalStateException("Client-id not provided")))
```
Should we do the same getOrElse for user above?

rajinisivaram replied:
The check verifies that if sanitizedClientId is provided, clientId is also provided. For user, we only have the sanitized version, so I think we don't need the check.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -421,86 +379,238 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
+  }
+
+  def shutdown() = {
+    throttledRequestReaper.shutdown()
```
Should we close the quotaCallback?

rajinisivaram replied:
We create one instance of `quotaCallback` that is shared across multiple quota managers. I had missed out the `close`, added one in `KafkaServer`.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -231,6 +233,17 @@ class KafkaApis(val requestChannel: RequestChannel,
+        val partitions = metadataCache.getAllPartitions()
+          .filter { case (_, state) => state.basePartitionState.leader != LeaderAndIsr.LeaderDuringDelete }
+          .map { case (tp, state) => (tp, TopicPartitionMetadata(state.basePartitionState.leader)) }
+          .toMap
```
.toMap is not needed.

junrao commented on core/src/test/scala/integration/kafka/api/BaseQuotaTest.scala:
```diff
@@ -100,154 +96,179 @@ abstract class BaseQuotaTest extends IntegrationTestHarness {
-      val throttleMetric = consumerThrottleMetric
-      throttled = throttleMetric != null && throttleMetric.value > 0
+      val metric = throttleMetric(ClientQuotaType.Fetch, consumerClientId)
+      throttled = metric != null && metric.value > 0
```
metric.value is deprecated. Could we use the new api? There are a few other places like that.

junrao commented on core/src/test/scala/integration/kafka/api/CustomQuotaCallbackTest.scala:
```diff
@@ -0,0 +1,385 @@
+  val DefaultFetchQuotaProp = "default.fetch.quota"
+}
+
+class GroupedUserQuotaCallback extends ClientQuotaCallback with Reconfigurable with Logging {
```
Could we add some doc of this class? For example, how is the naming convention of group for quota and topics.

junrao commented on core/src/test/scala/integration/kafka/api/CustomQuotaCallbackTest.scala:
```diff
@@ -0,0 +1,385 @@
+    user.produceConsume(expectProduceThrottle = true, expectConsumeThrottle = true)
+  }
+
+  private def createGroupWithOneUser(firstUser: String, brokerId: Int): GroupedUser = {
```
Perhaps add a comment on what brokerId is used for.

junrao commented on core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:
```diff
@@ -126,6 +127,35 @@ class DynamicBrokerConfigTest {
+  def testReconfigurableValidation(): Unit = {
+    val origProps = TestUtils.createBrokerConfig(0, TestUtils.MockZkConnect, port = 8181)
+    val config = KafkaConfig(origProps)
+    val invalidProps = Set(KafkaConfig.LogCleanerThreadsProp, KafkaConfig.BrokerIdProp, "some.prop")
```
"some.prop" is not a valid prop. So perhaps rename invalidProps to invalidReconfigurableProps ?

junrao reviewed (commented):
@rajinisivaram : Thanks for the patch. Left a few comments below.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@junrao Thanks for the review. I have moved the public interface to `clients` and converted it to Java. I have made most of the updates from the review, left a question about the interface.

rajinisivaram committed:
Address review comments

junrao commented on clients/src/main/java/org/apache/kafka/server/quota/ClientQuotaCallback.java:
```diff
@@ -0,0 +1,94 @@
+     * @param quotaType Type of quota requested
+     * @return quota metric tags that indicate which other clients share this quota
+     */
+    Map<String, String> quotaMetricTags(KafkaPrincipal principal, String clientId, ClientQuotaType quotaType);
```
Do we need quotaType? It seems that it's unrelated to metric tags.

rajinisivaram replied:
When multiple levels of quotas are used, the tags are dependent on which level config is chosen (e.g for {user, client} quota, both are included, for {user} quota, empty client-id is used). Since we allow configuring different levels of quota for Produce/Fetch/Request, they can potentially use different tags for the same client.

rajinisivaram replied:
Apart from `updateClusterMetadata`, all other methods are processed by individual `ClientQuotaManager`s and they pass in the quota type. I have moved the quota type as the first parameter of all these methods to keep them consistent.

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -360,56 +318,69 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
-        clientRateMetricName(clientQuotaEntity.sanitizedUser, clientQuotaEntity.clientId),
-        Some(getQuotaMetricConfig(clientQuotaEntity.quota)),
+        clientRateMetricName(metricTags),
+        Some(getQuotaMetricConfig(metricTags)),
```
This can be a bit tricky. If we call getQuotaMetricConfig() here, we will be paying the cost of calling ClientQuotaCallback.quotaLimit() on every request. We probably want to pass in a method and only call it when needed.

rajinisivaram replied:
I was thinking that the by-name parameter does lazy evaluation and hence the method doesn't get invoked on every request. I could be wrong, but I added a test and it seems to be true. What do you think?

junrao commented on core/src/main/scala/kafka/server/ClientQuotaManager.scala:
```diff
@@ -107,21 +158,26 @@ class ClientQuotaManager(private val config: ClientQuotaManagerConfig,
-    else QuotaTypes.ClientIdQuotaEnabled
+  private val clientQuotaType = quotaTypeToClientQuotaType(quotaType)
+  @volatile private var quotaTypesEnabled = clientQuotaCallback match {
+    case Some(_) => QuotaTypes.CustomQuotas
```
Is it useful to have CustomQuotas since it seems it's never used?

rajinisivaram replied:
It is used to avoid optimizing updates, the other values are used to check if only a single-level of quota is used. Have added a comment.

junrao commented on clients/src/main/java/org/apache/kafka/server/quota/ClientQuotaCallback.java:
```diff
@@ -0,0 +1,94 @@
+     * @param newValue    The new quota value
+     * @return true if quotas have changed and metric configs may need to be updated
+     */
+    boolean updateQuota(ClientQuotaEntity quotaEntity, ClientQuotaType quotaType, double newValue);
```
Is there a case when updateQuota() can return false? Ditto for removeQuota().

rajinisivaram replied:
I was using the return value with the default callback handler earlier, but now that it has been tidied up, we don't need it. Changed both `updateQuota` and `removeQuota` to `void` returns.

junrao commented on core/src/test/scala/unit/kafka/server/DynamicBrokerConfigTest.scala:
```diff
@@ -126,6 +127,35 @@ class DynamicBrokerConfigTest {
+    val origProps = TestUtils.createBrokerConfig(0, TestUtils.MockZkConnect, port = 8181)
+    val config = KafkaConfig(origProps)
+    val invalidReconfigurableProps = Set(KafkaConfig.LogCleanerThreadsProp, KafkaConfig.BrokerIdProp, "some.prop")
+    val validProps = Set(KafkaConfig.LogCleanerThreadsProp, KafkaConfig.LogCleanerDedupeBufferSizeProp, "some.prop")
```
We probably want to rename validProps to validReconfigurableProps for consistency?

junrao reviewed (commented):
@rajinisivaram : Thanks for the updated patch. A few more comments below.

rajinisivaram committed:
Address review comments

rajinisivaram commented:
@junrao Thanks for the review. I have updated the code and also made some changes to `CustomQuotaCallbackTest` to make it complete faster (the tests check only the metrics on the broker-side to avoid waiting for throttled requests to complete). Client-side metrics are verified by the other quota tests anyway.

I also added a new method to the callback interface to check if quotas need resetting. Since we now don't get quota limits from the callback handler on each request, we need to check if limits may have changed (e.g. with a callback that uses an external database to store quotas). Thoughts?

junrao reviewed (approved):
@rajinisivaram : Thanks for the updated patch. LGTM

rajinisivaram commented:
@junrao Thanks for the reviews, merging to trunk.
